# Mini Project2
### 20181531 김연주

In [6]:
import numpy as np
import pandas as pd

df = pd.read_csv("train.csv")
df

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
0              0.288585          -0.020294          -0.132905   
1              0.278419          -0.016411          -0.123520   
2              0.279653          -0.019467          -0.113462   
3              0.279174          -0.026201          -0.123283   
4              0.276629          -0.016570          -0.115362   
...                 ...                ...                ...   
7347           0.299665          -0.057193          -0.181233   
7348           0.273853          -0.007749          -0.147468   
7349           0.273387          -0.017011          -0.045022   
7350           0.289654          -0.018843          -0.158281   
7351           0.351503          -0.012423          -0.203867   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
0            -0.995279         -0.983111         -0.913526         -0.995112   
1            -0.998245         -0.975300         -0.960322         -0.998807   
2            -0.995380         -0.967187         -0.978944         -0.996520   
3            -0.996091         -0.983403         -0.990675         -0.997099   
4            -0.998139         -0.980817         -0.990482         -0.998321   
...                ...               ...               ...               ...   
7347         -0.195387          0.039905          0.077078         -0.282301   
7348         -0.235309          0.004816          0.059280         -0.322552   
7349         -0.218218         -0.103822          0.274533         -0.304515   
7350         -0.219139         -0.111412          0.268893         -0.310487   
7351         -0.269270         -0.087212          0.177404         -0.377404   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
0            -0.983185         -0.923527         -0.934724  ...   
1            -0.974914         -0.957686         -0.943068  ...   
2            -0.963668         -0.977469         -0.938692  ...   
3            -0.982750         -0.989302         -0.938692  ...   
4            -0.979672         -0.990441         -0.942469  ...   
...                ...               ...               ...  ...   
7347          0.043616          0.060410          0.210795  ...   
7348         -0.029456          0.080585          0.117440  ...   
7349         -0.098913          0.332584          0.043999  ...   
7350         -0.068200          0.319473          0.101702  ...   
7351         -0.038678          0.229430          0.269013  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                           -0.710304                    -0.112754   
1                           -0.861499                     0.053477   
2                           -0.760104                    -0.118559   
3                           -0.482845                    -0.036788   
4                           -0.699205                     0.123320   
...                               ...                          ...   
7347                        -0.880324                    -0.190437   
7348                        -0.680744                     0.064907   
7349                        -0.304029                     0.052806   
7350                        -0.344314                    -0.101360   
7351                        -0.740738                    -0.280088   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                                 0.030400                         -0.464761   
1                                -0.007435                         -0.732626   
2                                 0.177899                          0.100699   
3                                -0.012892                          0.640011   
4                                 0.122542                          0.693578   
...                                    ...                               ...   
7347                              0.829718                

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 31.6+ MB


In [7]:
from sklearn.utils import shuffle
df = shuffle(df)

In [8]:
from sklearn.preprocessing import LabelEncoder

df_cat = df['Activity']
encoder = LabelEncoder()
encoder.fit(df_cat)
df_encoded = encoder.transform(df_cat)

In [9]:
df['Activity'] = df_encoded
df['Activity']

2327    3
4630    0
14      2
7190    2
3161    0
       ..
2603    0
5425    4
2650    3
3132    0
6979    2
Name: Activity, Length: 7352, dtype: int32

In [10]:
X = df.drop(['subject', 'Activity'], axis=1)
y = df['Activity']

## ExtraTreesClassifier

In [7]:
from sklearn.ensemble import ExtraTreesClassifier

extra_clf = ExtraTreesClassifier(random_state=42)
extra_clf.fit(X, y)

ExtraTreesClassifier(random_state=42)

In [8]:
from sklearn.model_selection import cross_val_score

extra_score = cross_val_score(extra_clf, X, y, cv=10, scoring="accuracy") # 교차검증
extra_score

array([0.98097826, 0.98913043, 0.98911565, 0.98231293, 0.98639456,
       0.98367347, 0.98911565, 0.99319728, 0.99183673, 0.98231293])

In [9]:
extra_score.mean()

0.9868067879325645

## Models

In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

log_clf = LogisticRegression(solver="liblinear", random_state=42,max_iter=500)
rnd_clf = RandomForestClassifier(n_estimators=100, random_state=42)
svm_clf = SVC(kernel="linear", probability=True, random_state=42)
ada_clf = AdaBoostClassifier(RandomForestClassifier(random_state=42), n_estimators=150)
de_ada_clf = AdaBoostClassifier(DecisionTreeClassifier(random_state=42), n_estimators=150)
gbrt_clf = GradientBoostingClassifier(random_state=42)
bag_clf = BaggingClassifier(
    RandomForestClassifier(random_state=42), bootstrap=False, random_state=42)

In [14]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation", scores.std())

In [12]:
log_clf.fit(X, y)
rnd_clf.fit(X, y)
svm_clf.fit(X, y)
ada_clf.fit(X, y)
de_ada_clf.fit(X,y)
gbrt_clf.fit(X, y)
bag_clf.fit(X, y)

BaggingClassifier(base_estimator=RandomForestClassifier(random_state=42),
                  bootstrap=False, random_state=42)

In [13]:
log_score = cross_val_score(log_clf, X, y, cv=10, scoring="accuracy")
log_score

array([0.97146739, 0.98913043, 0.98231293, 0.98231293, 0.98095238,
       0.99047619, 0.99319728, 0.98367347, 0.99591837, 0.9877551 ])

In [14]:
display_scores(log_score)

Scores: [0.97146739 0.98913043 0.98231293 0.98231293 0.98095238 0.99047619
 0.99319728 0.98367347 0.99591837 0.9877551 ]
Mean: 0.9857196465542739
Standard deviation 0.006719985864114085


In [15]:
rnd_score = cross_val_score(rnd_clf, X, y, cv=10, scoring="accuracy")
rnd_score

array([0.97146739, 0.98369565, 0.98503401, 0.97823129, 0.98095238,
       0.97687075, 0.98095238, 0.98503401, 0.98911565, 0.98367347])

In [16]:
display_scores(rnd_score)

Scores: [0.97146739 0.98369565 0.98503401 0.97823129 0.98095238 0.97687075
 0.98095238 0.98503401 0.98911565 0.98367347]
Mean: 0.9815026989056491
Standard deviation 0.004752775899085641


In [17]:
svm_score = cross_val_score(svm_clf, X, y, cv=10, scoring="accuracy")
svm_score

array([0.98233696, 0.99048913, 0.98639456, 0.98367347, 0.98503401,
       0.99047619, 0.99047619, 0.98367347, 0.99455782, 0.98367347])

In [18]:
display_scores(svm_score)

Scores: [0.98233696 0.99048913 0.98639456 0.98367347 0.98503401 0.99047619
 0.99047619 0.98367347 0.99455782 0.98367347]
Mean: 0.9870785270629991
Standard deviation 0.003906071070233344


In [19]:
ada_clf_score = cross_val_score(ada_clf, X, y, cv=10, scoring="accuracy")
ada_clf_score

array([0.97282609, 0.98505435, 0.98231293, 0.97823129, 0.98639456,
       0.97687075, 0.97823129, 0.98367347, 0.99183673, 0.97687075])

In [20]:
display_scores(ada_clf_score)

Scores: [0.97282609 0.98505435 0.98231293 0.97823129 0.98639456 0.97687075
 0.97823129 0.98367347 0.99183673 0.97687075]
Mean: 0.9812302203490091
Standard deviation 0.005362469413583672


In [21]:
de_ada_clf_score = cross_val_score(de_ada_clf, X, y, cv=10, scoring="accuracy")
de_ada_clf_score

array([0.93342391, 0.94565217, 0.94421769, 0.94421769, 0.92653061,
       0.94829932, 0.94013605, 0.95238095, 0.94557823, 0.93333333])

In [22]:
display_scores(de_ada_clf_score)

Scores: [0.93342391 0.94565217 0.94421769 0.94421769 0.92653061 0.94829932
 0.94013605 0.95238095 0.94557823 0.93333333]
Mean: 0.9413769964507542
Standard deviation 0.007556317104098467


In [23]:
bag_clf_score = cross_val_score(bag_clf, X, y, cv=10, scoring="accuracy")
bag_clf_score

array([0.9701087 , 0.98913043, 0.98367347, 0.98095238, 0.98367347,
       0.97959184, 0.98231293, 0.98503401, 0.9877551 , 0.98095238])

In [24]:
display_scores(bag_clf_score)

Scores: [0.9701087  0.98913043 0.98367347 0.98095238 0.98367347 0.97959184
 0.98231293 0.98503401 0.9877551  0.98095238]
Mean: 0.9823184708666075
Standard deviation 0.004972440083167505


In [25]:
from sklearn.ensemble import ExtraTreesClassifier

extra_trees_clf = ExtraTreesClassifier(n_estimators=50, random_state=42)
extra_trees_clf.fit(X, y)

ExtraTreesClassifier(n_estimators=50, random_state=42)

In [26]:
from sklearn.model_selection import cross_val_score

extra_fold_score = cross_val_score(extra_trees_clf, X, y, cv=10, scoring="accuracy")
extra_fold_score

array([0.98097826, 0.98913043, 0.99183673, 0.97959184, 0.99047619,
       0.98231293, 0.98231293, 0.99455782, 0.99183673, 0.98231293])

In [27]:
display_scores(extra_fold_score)

Scores: [0.98097826 0.98913043 0.99183673 0.97959184 0.99047619 0.98231293
 0.98231293 0.99455782 0.99183673 0.98231293]
Mean: 0.9865346790890271
Standard deviation 0.005247993416948738


## Voting

In [28]:
voting_clf = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rnd_clf), ('svc', svm_clf)],
    voting='hard')

In [29]:
voting_clf.fit(X, y)

VotingClassifier(estimators=[('lr',
                              LogisticRegression(max_iter=500, random_state=42,
                                                 solver='liblinear')),
                             ('rf', RandomForestClassifier(random_state=42)),
                             ('svc',
                              SVC(kernel='linear', probability=True,
                                  random_state=42))])

In [30]:
from sklearn.model_selection import cross_val_score

voting_clf_score = cross_val_score(voting_clf, X, y, cv=10, scoring="accuracy")
voting_clf_score

array([0.97418478, 0.99184783, 0.98503401, 0.98367347, 0.98503401,
       0.99183673, 0.99319728, 0.98639456, 0.99591837, 0.98639456])

In [31]:
display_scores(voting_clf_score)

Scores: [0.97418478 0.99184783 0.98503401 0.98367347 0.98503401 0.99183673
 0.99319728 0.98639456 0.99591837 0.98639456]
Mean: 0.9873515601892929
Standard deviation 0.0058937419761207355


### Soft voting

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
log_clf = LogisticRegression(solver="lbfgs", random_state=42,max_iter=1000)
rnd_clf = RandomForestClassifier(n_estimators=100, random_state=42)
svm_clf = SVC(gamma="scale", probability=True, random_state=42)

softvoting_clf = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rnd_clf), ('svc', svm_clf)],
    voting='soft')
softvoting_clf.fit(X, y)

C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


VotingClassifier(estimators=[('lr',
                              LogisticRegression(max_iter=1000,
                                                 random_state=42)),
                             ('rf', RandomForestClassifier(random_state=42)),
                             ('svc', SVC(probability=True, random_state=42))],
                 voting='soft')

In [12]:
from sklearn.model_selection import cross_val_score

softvoting_clf_score = cross_val_score(softvoting_clf, X, y, cv=10, scoring="accuracy")
softvoting_clf_score

C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\user\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i 

array([0.98233696, 0.97418478, 0.97823129, 0.98911565, 0.98639456,
       0.98367347, 0.99047619, 0.9877551 , 0.97414966, 0.98503401])

In [15]:
display_scores(softvoting_clf_score)

Scores: [0.98233696 0.97418478 0.97823129 0.98911565 0.98639456 0.98367347
 0.99047619 0.9877551  0.97414966 0.98503401]
Mean: 0.9831351671103224
Standard deviation 0.0055772163146774285


- LogisticRegression: 0.9861265158237208
- RandomForestClassifier: 0.9823177314404022
- SVC: 0.988710255841467
- AdaBoostClassifier: 0.9805492088139603
- GradientBoostingClassifier:0.9865341245193731 
- BaggingClassifier: 0.9831340579710144

### SVM의 결과가 제일 좋게 나왔음

## 결과: 0.988710255841467 => 약 98.9%